File: sdk/node/metrics.js
/**
 * Node.js custom metrics template — all six OpenTelemetry instrument types.
 *
 * SETUP: auto.js must be loaded first. It initializes the global MeterProvider.
 * This file shows how to create and use every instrument type from that provider.
 *
 * REQUIRED ENV VARS (set before starting your app):
 *   OTEL_SERVICE_NAME             - e.g. "checkout-api"
 *   OTEL_EXPORTER_OTLP_ENDPOINT  - e.g. "http://localhost:4317"
 *
 * VERIFY IT'S WORKING:
 *   docker logs otel-collector --tail 30
 *   Query Prometheus: http://localhost:9090
 *   All metrics are prefixed with "otel_" by the Collector's prometheus exporter.
 *
 * INSTRUMENT TYPES — pick the right one:
 *
 *   Counter              Always goes up. Request counts, bytes sent, errors.
 *   UpDownCounter        Goes up and down. Active connections, queue depth, cache size.
 *   Histogram            Measures distributions. Latency, payload size, retry count.
 *   ObservableGauge      Snapshot of a value you poll. Memory usage, CPU %, config values.
 *   ObservableCounter    Monotonically increasing value you poll. Total GC collections, process uptime.
 *   ObservableUpDownCounter  Up/down value you poll. Thread pool size, open file handles.
 */

"use strict";

const { metrics } = require("@opentelemetry/api");

// One meter per logical component. The name becomes the instrumentation scope.
const meter = metrics.getMeter("com.example.checkout", "1.0.0");

// ─── Counter ──────────────────────────────────────────────────────────────────
// Use when: a value only ever increases and you want to track the rate of change.
// Prometheus query: rate(otel_orders_total[5m])
const ordersCounter = meter.createCounter("orders", {
  description: "Total orders submitted",
  unit: "1",
});

// Record with attributes — each unique attribute set is a separate Prometheus series.
// Keep attribute cardinality low: status and region are fine; user_id is not.
function recordOrder(status, region) {
  ordersCounter.add(1, { "order.status": status, "deployment.region": region });
}

// ─── UpDownCounter ────────────────────────────────────────────────────────────
// Use when: a value goes both up and down and you need the current level.
// Prometheus query: otel_active_checkouts
const activeCheckouts = meter.createUpDownCounter("active_checkouts", {
  description: "Number of checkout sessions currently in progress",
  unit: "1",
});

function startCheckout() { activeCheckouts.add(1); }
function endCheckout()   { activeCheckouts.add(-1); }

// ─── Histogram ────────────────────────────────────────────────────────────────
// Use when: you need latency percentiles, not just averages.
// Prometheus query: histogram_quantile(0.95, rate(otel_checkout_duration_milliseconds_bucket[5m]))
const checkoutDuration = meter.createHistogram("checkout.duration", {
  description: "End-to-end checkout latency in milliseconds",
  unit: "ms",
  // advice.explicitBucketBoundaries: define buckets tuned to your SLO thresholds.
  // Default buckets (0, 5, 10, 25 ... 10000ms) are often too coarse for HTTP latency.
  advice: {
    explicitBucketBoundaries: [10, 25, 50, 100, 200, 400, 800, 1600, 3200],
  },
});

async function timedCheckout(checkoutFn, method) {
  const start = performance.now();
  try {
    const result = await checkoutFn();
    checkoutDuration.record(performance.now() - start, {
      "payment.method": method,
      "checkout.status": "success",
    });
    return result;
  } catch (err) {
    checkoutDuration.record(performance.now() - start, {
      "payment.method": method,
      "checkout.status": "error",
    });
    throw err;
  }
}

// ─── ObservableGauge ──────────────────────────────────────────────────────────
// Use when: you want to report the current value of something at scrape time.
// The callback is called when the Collector reads metrics — don't do expensive work here.
// Prometheus query: otel_process_heap_used_bytes
meter.createObservableGauge("process.heap.used", {
  description: "V8 heap used in bytes",
  unit: "By",
}).addCallback((observableResult) => {
  const mem = process.memoryUsage();
  // Each observe() call with different attributes creates a separate time series.
  observableResult.observe(mem.heapUsed,  { "heap.space": "used" });
  observableResult.observe(mem.heapTotal, { "heap.space": "total" });
  observableResult.observe(mem.rss,       { "heap.space": "rss" });
});

// ─── ObservableCounter ────────────────────────────────────────────────────────
// Use when: the raw value is monotonically increasing but you read it by polling
// (e.g., from a system API, cache, or external source) rather than incrementing inline.
// Prometheus converts the reported absolute values to a rate automatically.
// Prometheus query: rate(otel_http_requests_handled_total[5m])
let _totalRequestsHandled = 0; // imagine this is updated by your request handler
meter.createObservableCounter("http.requests.handled", {
  description: "Total HTTP requests handled since process start",
  unit: "1",
}).addCallback((observableResult) => {
  observableResult.observe(_totalRequestsHandled);
});

// ─── ObservableUpDownCounter ──────────────────────────────────────────────────
// Use when: a pool or queue size is maintained externally and you poll it.
// Prometheus query: otel_db_pool_connections
meter.createObservableUpDownCounter("db.pool.connections", {
  description: "Database connection pool state",
  unit: "1",
}).addCallback((observableResult) => {
  // Replace getPoolStats() with your actual pool library's API.
  const stats = getPoolStats();
  observableResult.observe(stats.active, { "pool.state": "active" });
  observableResult.observe(stats.idle,   { "pool.state": "idle" });
  observableResult.observe(stats.waiting,{ "pool.state": "waiting" });
});

// ─── BatchObserver — observe multiple instruments in one callback ──────────────
// Efficient when you read from a single source (e.g. one syscall) to populate
// multiple instruments. All observations share the same timestamp.
const heapGauge   = meter.createObservableGauge("v8.heap.space.used",  { unit: "By" });
const externalGauge = meter.createObservableGauge("v8.external.memory", { unit: "By" });

meter.addBatchObservableCallback(
  (observableResult) => {
    const mem = process.memoryUsage();
    observableResult.observe(heapGauge,    mem.heapUsed);
    observableResult.observe(externalGauge, mem.external);
  },
  [heapGauge, externalGauge],
);

// ─── Placeholder helpers (replace with your real implementations) ──────────────
function getPoolStats() { return { active: 5, idle: 3, waiting: 0 }; }

module.exports = { recordOrder, startCheckout, endCheckout, timedCheckout };

File: sdk/python/metrics.py
"""
Python custom metrics template — all six OpenTelemetry instrument types.

SETUP: auto.py must be imported first. It initializes the global MeterProvider.
This module shows how to create and use every instrument type from that provider.

REQUIRED ENV VARS (set before starting your app):
  OTEL_SERVICE_NAME             - e.g. "checkout-api"
  OTEL_EXPORTER_OTLP_ENDPOINT  - e.g. "http://localhost:4317"

VERIFY IT'S WORKING:
  docker logs otel-collector --tail 30
  Query Prometheus: http://localhost:9090
  All metrics are prefixed with "otel_" by the Collector's prometheus exporter.

INSTRUMENT TYPES — pick the right one:

  Counter                  Always goes up. Request counts, bytes sent, errors.
  UpDownCounter            Goes up and down. Active connections, queue depth, cache size.
  Histogram                Measures distributions. Latency, payload size, retry count.
  ObservableGauge          Snapshot of a value you poll. Memory, CPU %, config flag.
  ObservableCounter        Monotonically increasing value you poll. Total GC runs, uptime.
  ObservableUpDownCounter  Up/down value you poll. Thread pool size, open file handles.
"""

import os
import time
import threading
from opentelemetry import metrics

# One meter per logical component. The name becomes the instrumentation scope.
meter = metrics.get_meter("com.example.checkout", "1.0.0")


# ─── Counter ──────────────────────────────────────────────────────────────────
# Use when: a value only ever increases and you want to track the rate of change.
# Prometheus query: rate(otel_orders_total[5m])
orders_counter = meter.create_counter(
    "orders",
    description="Total orders submitted",
    unit="1",
)

def record_order(status: str, region: str) -> None:
    # Attributes are the Prometheus label dimensions for this metric.
    # Keep cardinality low: "status" and "region" are fine; "user_id" is not.
    orders_counter.add(1, {"order.status": status, "deployment.region": region})


# ─── UpDownCounter ────────────────────────────────────────────────────────────
# Use when: a value goes both up and down and you need the current level.
# Prometheus query: otel_active_checkouts
active_checkouts = meter.create_up_down_counter(
    "active_checkouts",
    description="Number of checkout sessions currently in progress",
    unit="1",
)

def start_checkout() -> None: active_checkouts.add(1)
def end_checkout()   -> None: active_checkouts.add(-1)


# ─── Histogram ────────────────────────────────────────────────────────────────
# Use when: you need percentile visibility, not just averages.
# Prometheus query: histogram_quantile(0.95, rate(otel_checkout_duration_ms_bucket[5m]))
checkout_duration = meter.create_histogram(
    "checkout.duration",
    description="End-to-end checkout latency in milliseconds",
    unit="ms",
)

def timed_checkout(checkout_fn, method: str):
    start = time.perf_counter()
    try:
        result = checkout_fn()
        elapsed_ms = (time.perf_counter() - start) * 1000
        checkout_duration.record(elapsed_ms, {
            "payment.method": method,
            "checkout.status": "success",
        })
        return result
    except Exception:
        elapsed_ms = (time.perf_counter() - start) * 1000
        checkout_duration.record(elapsed_ms, {
            "payment.method": method,
            "checkout.status": "error",
        })
        raise


# ─── ObservableGauge ──────────────────────────────────────────────────────────
# Use when: you want to report the current value of something at collection time.
# The callback is invoked by the SDK on each export cycle — keep it fast.
# Prometheus query: otel_process_memory_rss_bytes
def _observe_process_memory(options: metrics.CallbackOptions):
    import resource
    rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    # On Linux ru_maxrss is in kilobytes; on macOS it's bytes.
    multiplier = 1 if os.uname().sysname == "Darwin" else 1024
    yield metrics.Observation(rss * multiplier, {"process": "self"})

meter.create_observable_gauge(
    "process.memory.rss",
    callbacks=[_observe_process_memory],
    description="Resident set size of the current process",
    unit="By",
)


# ─── ObservableCounter ────────────────────────────────────────────────────────
# Use when: a counter is maintained externally (e.g., in a library, cache, or
# C extension) and you can only read its current absolute value, not increment it.
# The SDK converts successive absolute values into a rate automatically.
# Prometheus query: rate(otel_http_requests_handled_total[5m])
_total_requests_handled = 0  # updated elsewhere in your application
_lock = threading.Lock()

def _observe_total_requests(options: metrics.CallbackOptions):
    with _lock:
        yield metrics.Observation(_total_requests_handled)

meter.create_observable_counter(
    "http.requests.handled",
    callbacks=[_observe_total_requests],
    description="Total HTTP requests handled since process start",
    unit="1",
)


# ─── ObservableUpDownCounter ──────────────────────────────────────────────────
# Use when: a pool, queue, or connection count is maintained by an external
# library and you need to poll it rather than instrument it inline.
# Prometheus query: otel_db_pool_connections
def _observe_db_pool(options: metrics.CallbackOptions):
    stats = _get_pool_stats()  # your DB pool library's status method
    yield metrics.Observation(stats["active"],  {"pool.state": "active"})
    yield metrics.Observation(stats["idle"],    {"pool.state": "idle"})
    yield metrics.Observation(stats["waiting"], {"pool.state": "waiting"})

meter.create_observable_up_down_counter(
    "db.pool.connections",
    callbacks=[_observe_db_pool],
    description="Database connection pool state by connection status",
    unit="1",
)


# ─── Multi-instrument batch callback ──────────────────────────────────────────
# Read from a single source once and populate multiple instruments.
# Avoids calling an expensive API (e.g., a syscall or HTTP call) separately
# for each instrument.
_heap_gauge      = meter.create_observable_gauge("python.gc.objects",  unit="1")
_gen_gauge       = meter.create_observable_gauge("python.gc.generation", unit="1")

def _observe_gc(options: metrics.CallbackOptions):
    import gc
    counts = gc.get_count()  # (gen0, gen1, gen2) — objects surviving each generation
    _heap_gauge.observe(sum(counts))
    for i, count in enumerate(counts):
        _gen_gauge.observe(count, {"gc.generation": str(i)})

# Note: batch callbacks via create_observable_* with a shared callback are the
# Python idiom — there is no separate BatchObservableCallback API in the Python SDK.


# ─── Placeholder helper (replace with your real pool library's API) ────────────
def _get_pool_stats() -> dict:
    return {"active": 5, "idle": 3, "waiting": 0}
